ha_mroonga.cpp

```diff
@@ -14954,55 +14954,121 @@ mrn_io_and_cpu_cost ha_mroonga::scan_time()
     MRN_SET_BASE_TABLE_KEY(this, table);
   } else {
     MRN_SET_WRAP_SHARE_KEY(share, table->s);
     MRN_SET_WRAP_TABLE_KEY(this, table);
-    res = wrap_handler->read_time(index, ranges, rows);
+    res = wrap_handler->MRN_HANDLER_KEYREAD_TIME(index, ranges, rows, blocks);
     MRN_SET_BASE_SHARE_KEY(share, table->s);
     MRN_SET_BASE_TABLE_KEY(this, table);
   }
   DBUG_RETURN(res);
 }
 #endif
 
-double ha_mroonga::storage_read_time(uint index, uint ranges, ha_rows rows)
+mrn_io_and_cpu_cost ha_mroonga::storage_read_time(uint index,
+                                                  uint ranges,
+                                                  ha_rows rows,
+                                                  ulonglong blocks)
 {
   MRN_DBUG_ENTER_METHOD();
-  double time = handler::read_time(index, ranges, rows);
+  mrn_io_and_cpu_cost time =
+    handler::MRN_HANDLER_KEYREAD_TIME(index, ranges, rows, blocks);
   DBUG_RETURN(time);
 }
 
+#ifdef MRN_HANDLER_USE_KEYREAD_TIME
+IO_AND_CPU_COST ha_mroonga::keyread_time(uint index,
+                                         uint ranges,
+                                         ha_rows rows,
+                                         ulonglong blocks)
+#else
 double ha_mroonga::read_time(uint index, uint ranges, ha_rows rows)
+#endif
+{
+  MRN_DBUG_ENTER_METHOD();
+  mrn_io_and_cpu_cost time;
+#ifdef MRN_ENABLE_WRAPPER_MODE
+  if (share->wrapper_mode)
+  {
+    time = wrapper_read_time(index, ranges, rows
+#  ifdef MRN_HANDLER_USE_KEYREAD_TIME
+                             , blocks
+#  endif
+                             );
+  } else {
+#endif
+    time = storage_read_time(index, ranges, rows
+#ifdef MRN_HANDLER_USE_KEYREAD_TIME
+                             , blocks
+#endif
+                             );
+#ifdef MRN_ENABLE_WRAPPER_MODE
+  }
+#endif
+  DBUG_RETURN(time);
+}
+
+#ifdef MRN_HANDLER_USE_KEYREAD_TIME
+#  ifdef MRN_ENABLE_WRAPPER_MODE
+IO_AND_CPU_COST ha_mroonga::wrapper_rnd_pos_time(ha_rows rows)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST res;
+  MRN_SET_WRAP_SHARE_KEY(share, table->s);
+  MRN_SET_WRAP_TABLE_KEY(this, table);
+  res = wrap_handler->rnd_pos_time(rows);
+  MRN_SET_BASE_SHARE_KEY(share, table->s);
+  MRN_SET_BASE_TABLE_KEY(this, table);
+  DBUG_RETURN(res);
+}
+#  endif
+
+IO_AND_CPU_COST ha_mroonga::storage_rnd_pos_time(ha_rows rows)
 {
   MRN_DBUG_ENTER_METHOD();
-  double time;
+  IO_AND_CPU_COST res;
+  res = handler::rnd_pos_time(rows);
+  DBUG_RETURN(res);
+}
+
+IO_AND_CPU_COST ha_mroonga::rnd_pos_time(ha_rows rows)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST time;
 #ifdef MRN_ENABLE_WRAPPER_MODE
   if (share->wrapper_mode)
   {
-    time = wrapper_read_time(index, ranges, rows);
+    time = wrapper_rnd_pos_time(rows);
   } else {
 #endif
-    time = storage_read_time(index, ranges, rows);
+    //    time = storage_rnd_pos_time(rows);
 #ifdef MRN_ENABLE_WRAPPER_MODE
   }
 #endif
   DBUG_RETURN(time);
 }
+#endif
```

Comment:
Which `#ifdef` corresponds to this `#endif?`

Reply:
How about adding the condition `ifdef MRN_HANDLER_USE_KEYREAD_TIME` to each function if it's possible ?

Reply:
> Which `#ifdef` corresponds to this `#endif?`

This `#endif` corresponds to https://github.com/mroonga/mroonga/pull/757/files#diff-d58ab8d0141ec15fa46f249f151b763f495c4d634647aca9c795a7b027ea263eR15032.

Reply:
> How about adding the condition `ifdef MRN_HANDLER_USE_KEYREAD_TIME` to each function if it's possible ?

We don't add `#ifdef ... #endif` to each function in the other codes.(e.g https://github.com/mroonga/mroonga/blob/main/ha_mroonga.cpp#L365-L450)
So, I think that I should same as the other codes in this modification.

Reply:
Ah I got it.